### PR TITLE
Reworking line numbers + other fixes

### PR DESCRIPTION
**Pull request by harshad1**
@gsantner @guang-lin

I reworked some of the line number stuff in this PR:

- Simplified how the line number calculations and drawing are done.
- Make the line number setting per-file

Please take a look and feel free to take anything you feel is useful.

### CODE REVIEW

File: app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java [hunk 1]
```diff
@@ -64,16 +61,11 @@ public class HighlightingEditor extends AppCompatEditText {
 
     // For drawing line numbers
     private final Paint _paint = new Paint();
-    private ScrollView _scrollView;
-    private int _x;
-    private int _maxLineNumber = 1;
-    private int _maxLineNumberWidth;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java]: Removed need for many vars

File: app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java [hunk 2]
```diff
@@ -88,38 +80,18 @@ public HighlightingEditor(Context context, AttributeSet attrs) {
...
-            private final Pattern pattern = Pattern.compile("\n");
-            private Matcher matcher;
-
-            @Override
-            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
-                if (after == 0 && count > 0) {
-                    CharSequence deleted = s.subSequence(start, start + count);
-                    matcher = pattern.matcher(deleted);
-                    while (matcher.find()) {
-                        _maxLineNumber--;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java]: Removed need for change tracking

File: app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java [hunk 3]
```diff
@@ -166,90 +127,82 @@ public boolean onPreDraw() {
...
-                            _firstVisibleLine[1] = number - 1;
-                            _firstVisibleLineUpdated = true;
-                        }
-                        canvas.drawText(String.valueOf(number), _x, y + offsetY, _paint);
-                    }
+        int line = 0;
+        int number = 1;
+        int yPos = top;
+
+        while(line < maxLayoutLine && yPos < bottom) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java:182]: Simplified loop. Basic logic is same though

**Comment by harshad1** [comment]
@guang-lin @gsantner This seems to be working fine except for line numbers in view mode. Line numbers in view mode appear to be broken in the original branch too though. Feel free to merge this branch if you prefer or pull changes from here into other branches if desired.

**Comment by guang-lin** [comment]
It seems that you improved a lot, I will take a look carefully.

**Comment by gsantner** [comment]
would be also OK to merge the version fron guang-lin to master and then just harshads branch on top

**Comment by harshad1** [comment]
That is perfectly fine by me. Note that this branch includes all of @guang-lin's commits

**Comment by harshad1** [comment]
Also note that I made some more changes to how we do the numbering this morning. It is somewhat more efficient / faster now. The earlier version it was possible to see some drops if you fast scrolled through a very large file

**Comment by gsantner** [comment]
Changed the branch to linenumbers branch. Merged the initial PR and improvements can be continued via additional PRs to linenumbers branch. Then we can merge it to master when we think it's fine.

For this PR tried rebasing, but there were quite a lot conflicts. However a merge-resolve seems to be only resolving conflicts in 3 files.

**Comment by gsantner** [comment]
These changes are already integrated into guang's branch right? So this one can be closed I guess.

I think you also have the possibility to directly commit to the guang PR branch via git@github.com:guang remote.

**Comment by harshad1** [comment]
Closing as all commits have been moved

**Pull request closed without merging**